content/post/fewer-parameters-means-fewer-bugs.md

```diff
@@ -0,0 +1,248 @@
+#### Old pattern without refactoring
+```rb
+def summary(first_name, last_name, middle_name = nil)
+  if middle_name.nil?
+    "#{title} - #{last_name}, #{first_name} (#{year_published})"
+  else
+    "#{title} - #{last_name}, #{middle_name}, #{first_name} (#{year_published})"
+  end
+end
+```
+#### After refactoring
+```rb
+def summary
+  "#{title} - #{author.full_name} (#{year_published})"
+end
+```
+Obviously, the code quality has a huge improvement after refactoring.
+
+## Loose coupling
+Before we finish refactoring and send off this PR to someone, let's check the code again. Is it clean enough?
+
+The code seems short and simple, but this method still depends on the author class. In other words, the `summary` method needs to know the author object has a `full_name` method, we could say the book class still couple with author class. In this case, maybe it's hard to notice the cost of coupling, let's imagine a few years later, the `summary` method might become more complicated:
+
+#### Few years later
+```rb
+def summary
+  # 10 lines of code you don't want to read.
+  middle_result = "#{some_author_title_helper(author)} - #{author.full_name}"
+  # another 10 lines of code
+  return final_result
+end
+```
+The disadvantage of coupling is that when one class needs to change, you need to change others as well. For example, one day we need to rename the full name method or add more parameters to it, we need to modify the book class at the same time, which we don't like.
+
+The bad news is that in the real world it's almost impossible to remove all the coupling between objects. But we can at least reduce the impact by segregating and highlighting the dependence. 
+
+```rb
+def summary
+  "#{title} - #{author_name} (#{year_published})"
+end
+
+def author_name
+  author.full_name
+end
+```
+After this refactoring, 100% of the calls from the `summary` method are to the book object itself. The `Book` class still has the coupling with the `Author` class, but the point of coupling is obvious and easy to maintain. If we only consider the example case, you might wonder: "Isn't that a little bit over-designed?" Well, I would say that a good pattern is never too early. Now we can push the PR confidently and enjoy the holiday.
+
+## Conclusion
+Good code should be small and simple. Of course, parameters are not the only place we can put our effort into when we're looking to reduce code size. During refactoring, don't forget to confirm the result by unit testing to make sure the final behavior is consistent. People around you may think that you're wasting your time because the final behavior remains the same even though you spent a lot of time on it. But code quality is a value unto itself. I believe this line of thinking would help you perform well in development speed and product stability in the long term.
```

Comment:
```suggestion
Good code should be small and simple. Of course, parameters are not the only place we can put our effort into when we're looking to reduce code size.

During refactoring, don't forget to confirm the result by unit testing to make sure the final behavior is consistent. People around you may think that you're wasting your time because the final behavior remains the same even though you spent a lot of time on it. But there is value in code quality[^2]. I believe this line of thinking would help you perform well in development speed and product stability in the long term.

[^2]: For example, the [Design-Stamina Hypothesis](https://www.martinfowler.com/bliki/DesignStaminaHypothesis.html) which postulates that good design outperform bad/no design over time, in terms of functional delivery.
```

I think we should split the first paragraph because those are 2 distinct ideas.

1. summary: good code = small & simple code
2. refactoring may be seen as a waste of time **BUT** we disagree

I added a footnote to the "there's value in design" statement to give it more of a foundation.